README.md
# gplsim

Generalized partially linear single-index models (GPLSIM) are important tools in nonparametric regression. They extend popular generalized linear models to allow flexible nonlinear dependence on some predictors while overcoming the “curse of dimensionality”.

In gplsim package, we provides functions that employ penalised spline (P-spline) to estimate generalised partially linear single index models, which extend the generalised linear models to include nonlinear effect for some predictors.

We are trying to post this package to CRAN for more general users.

Or you can install it simply with the following code in R:
```
devtools::install_github("zzz1990771/gplsim")
library(gplsim)
?gplsim
```

If you run into any dependency issues, you can install the dependent package manually:

```
install.packages("mgcv","minpack.lm")
```

Quick example:

```
# parameter settings
n=1000
true.theta = c(1, 1, 1)/sqrt(3)
# Generating data (binary case)
# This function generates a sin bump model as in Yu et al. (2007).
# You may use your data instead.
data <- generate_data(n,true.theta=true.theta,family="binomial")
y=data$Y       # binary response
X=data$X       # single index term ;
Z=data$Z       # partially linear term ;

# Fit the generalised partially linear single-index models
result <- gplsim(y,X,Z,family = binomial)

# Estimation of Theta
result$theta

# The coefficients of the fitted model. 
# Parametric coefficients are first, followed by coefficients for each spline term in turn.
result$coefficients

# summary of the fitted model
summary(result)

#plot the estimated single index function curve
plot_si(result)


# Gaussian case
# This function generate a plain sin bump model with gaussian response.
data <- generate_data(n,true.theta=true.theta,family="gaussian")
y=data$Y       # continous response
X=data$X       # single index term ;
Z=data$Z       # partially linear term ;

result <- gplsim(y,X,Z,family = gaussian)
result$theta
result$coefficients
summary(result)


#plot the estimated single index function curve
plot_si(result)
#par(new=T)
#plot_si(result,index=Z,xaxt="n", yaxt="n",col="red")

# A real data example
data(air)
y=air$ozone               # response
X=as.matrix(air[,3:4])    # single index term ;
Z=air[,2]                 # partially linear term ;

result <- gplsim(y,X,Z=Z,family = gaussian,k=10)
result$theta
result$coefficients
summary(result)

# Or you can try different spline basis
result <- gplsim(y,X,Z=Z,family = gaussian,bs="tp",k=10)
result$theta
result$coefficients
summary(result)

```


NEWS.md
# GPLSIM

# GPLSIM version 0.9.1

This is a hot fix for the function plot.si. It has been changed to plot_si as it is not a S3 method. And the augments are now handled well.

# GPLSIM version 0.9.2

We now include unit tests and formula methods!

We also added more supports on distributions, such as Gamma.

# GPLSIM version 0.9.3

Allow for large toleration when testing.


test/sim.R
library(doParallel)
library(foreach)
cl <- makeCluster(12)
registerDoParallel(cl)

# Gaussian case
# parameter settings
M=200
gaussian_coefs <- foreach(i=1:M, .combine=rbind,.packages = c("gplsim","splines","mgcv")) %dopar%{
  n=1000
  true.theta = c(1, 1, 1)/sqrt(3)
  # This function generate a plain sin bump model with gaussian response.
  data <- generate_data(n,true.theta=true.theta,family="Gaussian")
  y=data$Y       # continous response
  X=data$X       # single index term ;
  Z=data$Z       # partially linear term ;
  result <- gplsim(y,X,Z,user.init=NULL,family = gaussian)
  c(result$theta,result$gamma)
}


apply(gaussian_coefs,2,mean)
apply(gaussian_coefs,2,sd)



# Binomial case
# parameter settings
M=200
binomial_coefs <- foreach(i=1:M, .combine=rbind,.packages = c("gplsim","splines","mgcv")) %dopar%{
  n=1000
  true.theta = c(1, 1, 1)/sqrt(3)
  # This function generate a plain sin bump model with Binomial response.
  data <- generate_data(n,true.theta=true.theta,family="Binomial")
  y=data$Y       # Binomial response
  X=data$X       # single index term ;
  Z=data$Z       # partially linear term ;
  result <- gplsim(y,X,Z,user.init=true.theta,family = binomial)
  c(result$theta,result$gamma)
}



apply(binomial_coefs,2,mean)
apply(binomial_coefs,2,sd)



# poisson case
# parameter settings
M=200
poisson_coefs<- foreach(i=1:M, .combine=rbind,.packages = c("splines","mgcv","gplsim")) %dopar%{
  n=1000
  true.theta = c(1, 1, 1)/sqrt(3)
  # This function generate a plain sin bump model with gaussian response.
  data <- generate_data(n,true.theta=true.theta,family="Poisson")
  y=data$Y       # poisson response
  X=data$X       # single index term ;
  Z=data$Z       # partially linear term ;
  result <- gplsim(y,X,Z,user.init=NULL,family = poisson(link = "log"),k=13)
  c(result$theta,result$gamma)
}

apply(poisson_coefs,2,mean)
apply(poisson_coefs,2,sd)
stopCluster(cl)




test/test.R
#######################################
############     gplsim    ############


#######################################
n=10000
m=1
true.alpha = c(1, 1, 1)/sqrt(3)


# Generating data (binary)
data <- generate_data(n,true.alpha=true.alpha,binary=TRUE)
y=data$Y       # binary response
X=data$X   # single index term ;
Z=data$Z       # partially linear term ;

result <- gplsimPs(y,X,Z,user.init=c(0,0),family = binomial)
result$theta
result$coefficients
summary(result)

# Generating data (gaussian)
data <- generate_data(n,true.alpha=true.alpha,binary=FALSE)
y=data$Y       # binary response
X=data$X   # single index term ;
Z=data$Z       # partially linear term ;

result <- gplsimPs(y,X,Z,user.init=c(0,0),family = gaussian)
result$theta
result$coefficients
summary(result)








# with iteration;
para.old = rep(0, 3)
ptm <- proc.time()
while((delta>=epsilon)&step<100){
  step=step+1

  alpha.old=para.old[1:3]
  er <- optim(alpha.old, si, y=y2,x=X,z=Z,fam=gaussian,hessian=TRUE,fx=TRUE,k=10)
  er <- optim(er$par, si, y=y2,x=X,z=Z,fam=gaussian,hessian=TRUE,k=10)
  b <- si(er$par, y2, X, Z, fam=gaussian, opt=FALSE)    ## best fit model
  alpha=b$theta                 # /sqrt(sum((b$theta)^2)) ;
  gamma=b$coefficients[1]
  para=c(alpha, gamma)
  delta = max(abs(para-para.old))
  para.old = para
}
proc.time() - ptm
#######################


###  Modified ;
### use standardized coefficient from gam
coeff = gam(y~X+Z-1, family=gaussian)$coefficients
alpha0 <- coeff[1:3]*sign(coeff[1])/sqrt(sum(coeff[1:3]^2))
## get initial alpha, using no penalization...
er <- optim(alpha0, si2, y=y,x=X,z=Z,fam=gaussian, hessian=TRUE,fx=TRUE,k=10)
#er$par
## now get alpha with smoothing parameter selection...
er <- optim(er$par, si2, y=y, x=X,z=Z,fam=gaussian,hessian=TRUE,k=10)
b <- si2(er$par, y, X, Z, fam=gaussian, opt=FALSE) ## best fit model
b$theta
b$coefficients[1:2]


#binomial
coeff = gam(y~X+Z-1, family=binomial )$coefficients
alpha0 <- coeff[1:3]*sign(coeff[1])/sqrt(sum(coeff[1:3]^2))
## get initial alpha, using no penalization...
er <- optim(alpha0, si2, y=y,x=X,z=Z,fam=binomial , hessian=TRUE,fx=TRUE,k=10)
#er$par
## now get alpha with smoothing parameter selection...
er <- optim(er$par, si2, y=y, x=X,z=Z,fam=binomial ,hessian=TRUE,k=10)
b <- si2(er$par, y, X, Z, fam=binomial , opt=FALSE) ## best fit model
b$theta
b$coefficients[1:2]


# try iteration;
coeff = gam(y~X+Z-1, family=binomial)$coefficients
alpha0 <- coeff[1:3]*sign(coeff[1])/sqrt(sum(coeff[1:3]^2))
epsilon = 10^(-4)
delta =1
step=0

para.old = c(alpha0, coeff[4])
ptm <- proc.time()
while((delta>=epsilon)&step<100){
  step=step+1

  alpha.old=para.old[1:3]
  er <- optim(alpha.old, si2, y=y,x=X,z=Z,fam=gaussian, hessian=TRUE,fx=TRUE,k=10)
  er <- optim(er$par, si2, y=y,x=X,z=Z,fam=gaussian, hessian=TRUE,k=10)
  b <- si2(er$par, y,X,Z, fam=gaussian, opt=FALSE)    ## best fit model
  alpha=b$theta                 # /sqrt(sum((b$theta)^2)) ;
  gamma=b$coefficients[1]
  para=c(alpha, gamma)
  delta = max(abs(para-para.old))
  para.old = para
}
proc.time() - ptm
step
para





##########  case 1    ##########
tau0 = 0.5

nknots.2n205=NULL; min_bic.2n205=NULL; nd.2=NULL;
bic.2n205=NULL
an205 <-matrix(rep(0, 15), ncol=3)
bn205<-NULL
iter2.2n205 <- matrix(rep(0, 500), ncol=5);   step2n205<-NULL
fitn205 <- matrix(rep(0, 20), ncol=4)
para_estn205<-matrix(rep(0, 400), ncol=4)
epsilon=10^(-5);

ptm <- proc.time()
for (j in 1:5){

  # Generating data
  data <- generate_data(n,true.alpha=true.alpha)
  y=data$Y       # binary response
  X=data$X   # single index term ;
  Z=data$Z       # partially linear term ;

  for (i in 1:5){
    nd.2[i] = i

    para.old <- quantreg::rq(y~ X+Z-1, tau=tau0)$coefficients
    delta=1
    step=0
    while ((delta>=epsilon)&(step<1000)){
      step=step+1;
      alpha.old <-para.old[1:3]
      alpha.old=sign(alpha.old[1])*alpha.old/sqrt(crossprod(alpha.old))
      beta.old <-para.old[4]

      index = X%*%alpha.old
      mm=min(index)
      MM=max(index)
      seq=seq(mm, MM, by=(MM-mm)/nd.2[i])
      inter= seq[-(nd.2[i]+1)]
      kts=c(mm, mm, mm, inter, MM, MM, MM, MM)
      bs2 = splines::splineDesign(knots=kts, ord=4, x=index, derivs=rep(0, n), outer.ok = F)
      ys <- y - Z*beta.old
      theta2 <- quantreg::rq(ys ~ bs2-1, tau=tau0)$coefficients
      eta = bs2%*%theta2
      r = y - eta - Z*beta.old;
      bs.dev2 = splines::splineDesign(knots=kts, ord=4, x=index, derivs=rep(1, n), outer.ok=F)
      etada = (as.vector(bs.dev2%*%theta2))*X

      # etadb = -(bs2%*%solve(t(bs2)%*%bs2)%*%t(bs2))%*%Z
      # etad = cbind(etada, etadb)
      # f1 = etad;
      # f1[, 4] = f1[, 4] + Z;
      ##f1=f1*diag(std(f1).^(-1));
      #y_new = r + f1%*%para.old;
      #x_new = f1
      #para = rq(y_new ~ x_new-1, tau=tau0)$coefficients
      ys_simple = y - eta + etada%*%alpha.old
      xs_simple = cbind(etada, Z)
      para = rq(ys_simple ~ xs_simple - 1, tau=tau0)$coefficients

      alpha =sign(para[1])*para[1:3]/ifelse(sqrt(crossprod(para[1:3]))==0,0.1^8,sqrt(crossprod(para[1:3])))
      beta = para[4]
      para=c(alpha, beta)
      delta = max(abs(para-para.old))
      para.old = para
    }     # end of while loop

    step2n205[i]=step
    fitn205[i, ]=para
    an205[i, ]=fitn205[i, 1:3]
    bn205[i]=fitn205[i, 4]
    index = X%*%an205[i, ]
    mm=min(index)
    MM=max(index)
    seq=seq(mm, MM, by=(MM-mm)/nd.2[i])
    inter= seq[-(nd.2[i]+1)]
    kts=c(mm, mm, mm, inter, MM, MM, MM, MM)
    bs2 = splines::splineDesign(knots=kts, ord=4, x=index, derivs=rep(0, n), outer.ok = F)
    ys = y - Z*bn205[i]
    theta2 <- rq(ys~bs2-1, tau=tau0)$coefficients

    s = y - bs2%*%theta2 - Z*bn205[i]
    rho= tau0*s - s*(s<0)
    loss = mean(rho)
    bic.2n205[i] = log(loss)+log(n)*(nd.2[i]-1+4+1)/(2*n)
  }    # end of iteration over num of knots

  iter2.2n205[j, ] = step2n205
  nknots.2n205[j] = nd.2[which(bic.2n205==min(bic.2n205))]-1
  min_bic.2n205[j] = bic.2n205[which(bic.2n205==min(bic.2n205))]
  para_estn205[j, ] = fitn205[which(bic.2n205==min(bic.2n205)), ]
}
proc.time() - ptm

length(iter2.2t205[iter2.2t205>50])
summary(min_bic.2t205)
table(nknots.2t205)
apply(para_estt205, 2, mean)
apply(para_estt205, 2, var)
c(3, 2, 1)/sqrt(14)
loss = exp(min_bic.2n205 - log(n)*(nd.2[i]-1+4+1)/(2*n) )


test/test2.R
data(air)
y=air$ozone               # response
X=as.matrix(air[,3:4])    # single index term ;
Z=air[,2]                 # partially linear term ;
#Z=rep(0,length(Z))

result <- gplsim(y,X,Z,user.init=NULL,k=20,family = gaussian)
result$theta
result$coefficients
summary(result)

sort_index <- order(X%*%true.theta)
lines((X%*%true.theta)[sort_index],data$single_index_values[sort_index])

lines((X%*%result$theta)[sort_index],result$fitted.values[sort_index])

plot((X%*%result$theta)[sort_index],result$fitted.values[sort_index])


result$model$a
attr(result$smooth[[1]],"offset") <-  result$coefficients[1]
result$smooth[[1]]@offset
mgcv::plot.gam(result,xlab = "single index",ylab = "mean",select=1)

result$offset <- rep(result$coefficients[1],n)
mgcv::plot.gam(result)
result$coefficients[1]
# parameter settings
n=200
true.alpha = c(1, 1, 1)/sqrt(3)

# Gaussian case
# This function generate a plain sin bump model with gaussian response.
data <- generate_data(n,true.alpha=true.alpha,family="Gaussian")
y=data$Y       # continous response
X=data$X       # single index term ;
Z=data$Z       # partially linear term ;

result <- gplsim(y,X,Z,user.init=FALSE,family = gaussian)
result$theta
result$coefficients
summary(result)


#plot the estimated single index function curve
plot.si(result)
#par(new=T)
#plot.si(result,index=Z,xaxt="n", yaxt="n",col="red")



R/gplsim0.r
generate_data <- function(n,true.alpha=c(1, 1, 1)/sqrt(3)){
  X = matrix(runif(length(true.alpha)*n), ncol=length(true.alpha))
  U = X%*%true.alpha

  #Z<-rep(c(0,1),n/2)
  Z<-rnorm(n)
  q <- NULL
  py <- NULL
  for(i in 1:n){
    q[i] = sin( (U[i]-0.3912)*pi/(1.3409 -0.3912) ) + 0.3*Z[i]
    py[i] = exp(q[i])/(1+exp(q[i]))
  }

  y= rbinom(n, size=1, prob=py)
  return(list("X" = X, "Y" = y,"Z"=Z))
}


si <- function(alpha,y,x,z,opt=TRUE,k=10,fam, fx=FALSE) {
  ## Fit single index model using gam call. Return ML is opt==TRUE
  ## and fitted gam with theta added otherwise...

  #reparameterized theta
  alpha_add1 <- c(1, alpha)
  theta <- alpha_add1/sqrt(sum(alpha_add1^2))

  a <- x%*%theta
  b <- mgcv::gam(y~s(a,fx=fx,k=k)+z-1,family=fam,method="ML")
  if (opt) return(b$gcv.ubre) else {
    b$theta <- theta  ## add theta
    return(b)
  }
}      # end of defining si;

si2 <- function(alpha,y,x,z,opt=TRUE,k=10,fam, fx=FALSE) {
  ## Fit single index model using gam call. Return ML is opt==TRUE
  ## and fitted gam with theta added otherwise...
  theta <- alpha
  theta <- theta*sign(theta[1])/sqrt(sum(theta^2))
  a <- x%*%theta
  b <- gam(y~s(a,fx=fx,k=k)+z,family=fam,method="ML")
  if (opt) return(b$gcv.ubre) else {
    b$theta <- theta  ## add theta
    return(b) }
}      # end of defining si;



gplsim<- gplsimPs <- function(Y=Y,X=X,Z=Z,family=binomial,penalty=TRUE,user.init=FALSE){
  p <- dim(X)[2]
  if(identical(FALSE,user.init)){
    if(length(user.init)!=(p-1)){
      stop("user.init length must be p-1")
    }else{
      init.alpha <- user.init
    }
  }else{
    er_np <- optim(rep(0,p-1), si, y=Y,x=X,z=Z, fam=family, hessian=TRUE, fx=TRUE)
    init.alpha <- er_np$par
  }

  er <- optim(init.alpha,si,y=Y,x=X,z=Z,fam=family,hessian=TRUE,fx=!penalty)
  b <- si(er_np$par,y=y,X,Z, fam=family, opt=FALSE)
  return(b)
}


R/gplsim.r
#' Data generation function for simulation and demonstration
#' A sine-bump setting has been employed.
#'
#' @param n sample size
#' @param true.theta true single-index coefficients,
#' default is c(1,1,1)/sqrt(3) for setting 1 and c(1,2)/sqrt(5) for other settings
#' @param family chose from "gaussian", "binomial" or "poisson".
#' @param ncopy generates multiple copies of data for Monte Carlo simulations
#'
#' @return X single index predictors
#' @return Y response variables, a list
#' @return Z partial linear predictor(s)
#' @return single_index_values single index term
#' @export
generate_data <- function(n,true.theta=c(1, 1, 1)/sqrt(3),family="gaussian",ncopy=1){
  #parameter setting
  sigma = 0.1
  c1 = 0.3912
  c2 = 1.3409
  rho = 0.3

  X = matrix(stats::runif(length(true.theta)*n), ncol=length(true.theta))
  U = X%*%true.theta
  fU = sin( (U-c1)*pi/(c2 -c1) )

  #Z<-rnorm(n)
  Z <- 1 - c(1:n)%%2
  q = as.vector(fU + rho*Z)

  if(family=="gaussian"){
    ylist <- lapply(vector(mode = "list", length = ncopy),function(x){q + rnorm(length(q),0,sigma)})
  }else if(family=="binomial"){
    py = exp(q)/(1+exp(q))
    ylist <- lapply(vector(mode = "list", length = ncopy),function(x){rbinom(length(q), size=1, prob=py)})
  }else if(family=="poisson"){
    py = exp(q)
    ylist <- lapply(vector(mode = "list", length = ncopy),function(x){rpois(length(q),py)})
  }else if(family=="gamma"){
    ylist <- lapply(vector(mode = "list", length = ncopy),function(x){
      mu <- q +1
      alpha <- exp(q)
      rgamma(length(q), shape=alpha, scale=mu/alpha)
    })

  }

  if(ncopy==1){ylist = ylist[[1]]}
  return(list("X" = X, "Y" = ylist,"Z"=Z,"single_index_values"=fU))
}

#' An internal function to optimization and fitting.
#' Don't use it solely.
#'
#' @param alpha single-index coefficients
#' @param y Response variable, should be a vector.
#' @param x Single index covariates.
#' @param z Partially linear covariates.
#' @param opt see ?gplsim
#' @param smooth_selection see ?gplsim
#' @param fam see ?gplsim
#' @param bs see ?gplsim
#' @param fx see ?gplsim
#' @param scale see ?gplsim
#' @param ... includes optional arguments user can pass to \code{mgcv::gam} or \code{glm}, such as \code{k}, which is the dimension of the basis of the smooth term and \code{m}, which is the order of the penalty for the smooth term
#'
#' @return b fitted gam object
si <- function(alpha,y,x,z,opt=TRUE,smooth_selection,fam,bs="ps", fx=FALSE,scale=scale,...) {
  ## from ..., expose to environment
  #print((...))
  list2env(list(...),environment())

  ## Fit single index model using gam call. Return ML is opt==TRUE
  ## and fitted gam with theta added otherwise...

  #reparameterized theta
  theta <- sign(alpha[1])*alpha/sqrt(sum(alpha^2))

  a <- x%*%theta
  if(is.null(z)){
    b <- mgcv::gam(y~s(a,bs=bs,fx=fx,m=m,k=k),family=fam,method= smooth_selection,scale=scale )
  }else{
    b <- mgcv::gam(y~s(a,bs=bs,fx=fx,m=m,k=k)+z,family=fam,method= smooth_selection,scale=scale )
    b$gamma <- b$coefficients[c(2:(1+NCOL(z)))]
  }
  if (opt) return(b$deviance) else {
    b$theta <- theta  ## add theta
    #b$gamma <- b$coefficients[c(2:(1+NCOL(z)))]
    class(b) <- c("gplsim","gam","glm","lm")
    return(b)
  }

}


#' @name gplsim
#' @export
gplsim <- function( ...)
  UseMethod("gplsim")





#' Function to fit generalized partially linear single-index models via penalized splines
#'
#' This function employs penalized spline (P-spline) to estimate generalized
#' partially linear single index models, which extend the generalized linear
#' models to include nonlinear effect for some predictors.
#'
#' For formula, method, see ?gplsim.formula
#'
#' @param Y Response variable, should be a vector.
#' @param X Single index covariates.
#' @param Z Partially linear covariates.
#' @param family A \code{family} object: a list of functions and expressions
#' for defining \code{link} and \code{variance} functions.
#' Families supported are \code{binomial}, \code{gaussian}.
#' The default family is \code{gaussian}.
#' @param penalty Whether use penalized splines or un-penalized splines to fit the model. The default is TRUE.
#' @param penalty_type The optional argument penalty_type is a character variable, which specifies the type of penalty used in the penalized splines estimation. The default penalty type is {L}_{2} penalty, while {L}_{1} is also supported.
#' @param profile profile is a logical variable that indicates whether the algorithm with profile likelihood or algorithm with NLS procedure should be used. The default algorithm is set to algorithm with profile likelihood.
#' @param bs bs is a character variable that specifies the spline basis in the estimation of unknown univariate function of single index. Default is P-splines.
#' @param user.init The user.init is a numeric vector of the same length as the dimensionality of single index predictors. The users can use this argument to pass in any appropriate user-defined initial single-index coefficients based on prior information or domain knowledge. The default value is NULL.
#' @param ... includes optional arguments user can pass to \code{mgcv::gam} or \code{glm}, such as \code{k}, which is the dimension of the basis of the smooth term and \code{m}, which is the order of the penalty for the smooth term. Others include:
#' \code{scale} The optional argument scale is a numeric indicator with a default value set to -1. Any negative value including -1 indicates that the scale of response distribution is unknown, thus need to be estimated. Another option is 0 signaling scale of 1 for Poisson and binomial distribution and unknown for others. Any positive value will be taken as the known scale parameter.
#' \code{smooth_selection} The optional argument smooth_selection is another character variable that specifies the criterion used in the selection of a smoothing parameter. The supported criteria include "GCV.Cp","GACV.Cp", "ML","P-ML", "P-REML" and "REML", while the default criterion is "GCV.Cp".
#'
#' @return theta  Estimation of Theta
#' @return coefficients  the coefficients of the fitted model. Parametric coefficients are first, followed by coefficients for each spline term in turn.
#' @return ... See GAM object
#' @examples
#' # parameter settings
#' n=200
#' true.theta = c(1, 1, 1)/sqrt(3)
#' # Gaussian case
#' # This function generate a plain sin bump model with gaussian response.
#' data <- generate_data(n,true.theta=true.theta,family="gaussian")
#' y=data$Y       # continous response
#' X=data$X       # single index term ;
#' Z=data$Z       # partially linear term ;
#'
#' result <- gplsim(y,X,Z,family = gaussian)
#' result$theta
#' result$coefficients
#' summary(result)
#'
#'
#' #plot the estimated single index function curve
#' plot_si(result)
#' @rdname gplsim
#' @method gplsim default
#' @export
gplsim.default <- function(Y=Y,X=X,Z=Z,family=gaussian(),penalty=TRUE,
                   penalty_type = "L2",profile = TRUE, bs="ps",user.init= NULL,...){
  #validate inputs of X,Y,Z
  stopifnot(is.numeric(Y),is.vector(Y))
  stopifnot(is.numeric(X),is.matrix(X))
  stopifnot("SIM predictos must no less than two"= dim(X)[2] >=2)
  stopifnot(is.numeric(X),is.matrix(X))
  stopifnot(is.null(Z) || (is.numeric(Z)) )

  #deal with ...
  args_list <- list(...)
  if (is.null(args_list$m)) args_list$m <- m <- 2
  if (is.null(args_list$k)) args_list$k <- k <- 13
  if (is.null(args_list$scale)) args_list$scale <- scale <- -1
  if (is.null(args_list$smooth_selection)) args_list$smooth_selection <- smooth_selection <- "GCV.Cp"
  list2env(list(...),environment())

  #check other arguments
  if(!is.null(user.init)){
      stopifnot("user.init length must be the same with X" = length(user.init) ==NCOL(X) )
      init.alpha <- user.init
  }else{
    if(is.null(Z)){
      linear <- glm(Y~X-1,family=family)
    } else{
      temp_XZ <- cbind(X,Z)
      linear <- glm(Y~temp_XZ-1,family=family)
    }
    init.alpha <- linear$coefficients[1:(NCOL(X))]
    init.alpha <- sign(init.alpha[1])*init.alpha/sqrt(sum(init.alpha^2))
  }


  #non-profile
  #test
  if(profile==FALSE){
    non.profile.fit <- plsiml(x=X,y=Y ,z=Z,degree=m,nknots=k,maxiter = 10,alpha0 = init.alpha)
    a <- X%*%non.profile.fit$alpha
    if(is.null(Z)){
      b <- mgcv::gam(Y~s(a,bs=bs,fx=!penalty,m=m,k=k),family=family,method= smooth_selection,scale=scale)
    }else{
      b <- mgcv::gam(Y~s(a,bs=bs,fx=!penalty,m=m,k=k)+Z,family=family,method= smooth_selection,scale=scale)
      b$gamma <- b$coefficients[c(2:(1+NCOL(Z)))]
    }
    b$theta <- non.profile.fit$alpha  ## add theta

    class(b) <- c("gplsim","gam","glm","lm")

  }else{
    er <- suppressWarnings(optim(init.alpha,si,y=Y,x=X,z=Z,fam=family, smooth_selection=smooth_selection, hessian=TRUE,fx=!penalty,bs=bs,scale=scale,m=m,k=k))
    b <- si(er$par,y=Y,X,Z,fam=family,smooth_selection=smooth_selection, opt=FALSE,bs=bs,scale=scale,m=m,k=k)
  }
  b$Xnames <- if(is.null(colnames(X)))  paste0("X.",seq(1,NCOL(X),by = 1)) else colnames(X)
  b$Znames <- if(is.null(colnames(Z)) && !is.null(Z))  paste0("Z.",seq(1,NCOL(Z),by = 1)) else colnames(Z)

  return(b)
}

#utils::globalVariables(c("yscale"))


#' Formula interface for gplsim
#'
#' This function add formula interface to gplsim function
#'
#' @param formula A model formula;
#' @param data A data matrix containing the variables in the formula.
#' @param family A \code{family} object: a list of functions and expressions
#' for defining \code{link} and \code{variance} functions.
#' Families supported are \code{binomial}, \code{gaussian}.
#' The default family is \code{gaussian}.
#' @param penalty Whether use penalized splines or un-penalized splines to fit the model. The default is TRUE.
#' @param penalty_type The optional argument penalty_type is a character variable, which specifies the type of penalty used in the penalized splines estimation. The default penalty type is {L}_{2} penalty, while {L}_{1} is also supported.
#' @param profile profile is a logical variable that indicates whether the algorithm with profile likelihood or algorithm with NLS procedure should be used. The default algorithm is set to algorithm with profile likelihood.
#' @param bs bs is a character variable that specifies the spline basis in the estimation of unknown univariate function of single index. Default is P-splines.
#' @param user.init The user.init is a numeric vector of the same length as the dimensionality of single index predictors. The users can use this argument to pass in any appropriate user-defined initial single-index coefficients based on prior information or domain knowledge. The default value is NULL.
#' @param ... includes optional arguments user can pass to \code{mgcv::gam} or \code{glm}, such as \code{k}, which is the dimension of the basis of the smooth term and \code{m}, which is the order of the penalty for the smooth term. Others include:
#' \code{scale} The optional argument scale is a numeric indicator with a default value set to -1. Any negative value including -1 indicates that the scale of response distribution is unknown, thus need to be estimated. Another option is 0 signaling scale of 1 for Poisson and binomial distribution and unknown for others. Any positive value will be taken as the known scale parameter.
#' \code{smooth_selection} The optional argument smooth_selection is another character variable that specifies the criterion used in the selection of a smoothing parameter. The supported criteria include "GCV.Cp","GACV.Cp", "ML","P-ML", "P-REML" and "REML", while the default criterion is "GCV.Cp".
#'
#' @return theta  Estimation of Theta
#' @return coefficients  the coefficients of the fitted model. Parametric coefficients are first, followed by coefficients for each spline term in turn.
#' @return ... See GAM object
#' @rdname gplsim
#' @method gplsim formula
#' @export
gplsim.formula <- function(formula, data, family=gaussian(), penalty=TRUE,
                   penalty_type = "L2",profile = TRUE, bs="ps",user.init= NULL,...){
  #parse formula
  lhs <- if(length(formula) == 3) formula[[2]] else NULL
  lhsVars <- all.vars(lhs)
  Y <- data[,lhsVars]

  rhs <- paste(deparse(formula[[length(formula)]], width.cutoff = 500), collapse="")
  rhs <- gsub("[[:blank:]]", "", rhs)
  si_vars <- strsplit(gsub("si\\(([^\\)]+)\\).*","\\1", rhs),split = "\\+")
  si_vars <- unlist(si_vars)
  X <- data[,si_vars]

  pl_vars <- strsplit(gsub("si\\(([^\\)]+)\\)(.*)","\\2", rhs),split = "\\+")
  pl_vars <- unlist(pl_vars)[-1]
  Z <- data[,pl_vars]

  cl <- match.call()
  cl[[1]] <- quote(gplsim)
  obj <- gplsim(Y=Y,X=X,Z=Z, ...)
  obj$call <- cl
  obj

}
